### PR TITLE
[Feature] Add DAFL Distillation

### PR DESCRIPTION
**Pull request by wilxy**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Add a new pipeline (DataFree Distillation) for Knowledge Distillation.

## Modification

1.Add a **DataFree Distillation Algorithm** named **DAFL**, including its config, Loss and readme.
2.Add `DataFreeDistillationtillation` for common **DataFree Distillation Algorithm** and `DAFLDataFreeDistillationtillation` for **DAFL**.
3.Add **Generators** used in **DataFree Distillation Algorithm**, including `base_generator` and `dafl_generator`.
4.Add `get_module_device` and `set_requires_grad functions` in utils.
5.Add UT for DAFLLoss.

## TODO

- [x] add UT for generators and DataFreeDistillationtillation.
- [x] add typing for all files

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmrazor/models/losses/dafl_loss.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        assert norm_type in ['norm', 'abs'], \
+            '"norm_type" must be "norm" or "abs"'
+        self.norm_type = norm_type
+
+        if self.norm_type == 'norm':
+            self.norm_fn = lambda x: -x.norm()
+        elif self.norm_type == 'abs':
+            self.norm_fn = lambda x: -x.abs().mean()
+
+    def forward(self, feat_T: torch.Tensor) -> torch.Tensor:
```
Comment (FreakieHuang) [inline mmrazor/models/losses/dafl_loss.py:62]: Is it exactly same func of its base class?

File: mmrazor/models/algorithms/distill/configurable/datafree_distillation.py [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+from mmcv.runner import load_checkpoint
+from mmengine.optim import OptimWrapper
+
+from mmrazor.models.utils import add_prefix, set_requires_grad
+from mmrazor.registry import MODELS
+from ...base import BaseAlgorithm
+
+
+@MODELS.register_module()
+class DataFreeDistillation(BaseAlgorithm):
```
Comment (FreakieHuang) [inline mmrazor/models/algorithms/distill/configurable/datafree_distillation.py:15]: UT

File: mmrazor/models/ops/gather_tensors.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from typing import Any
+
+import torch
+import torch.distributed as dist
+
+
+class GatherTensors(torch.autograd.Function):
```
Comment (FreakieHuang) [inline mmrazor/models/ops/gather_tensors.py:8]: UT

File: mmrazor/models/architectures/generators/dafl_generator.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from mmrazor.registry import MODELS
+from .base_generator import BaseGenerator
+
+
+@MODELS.register_module()
+class DAFLGenerator(BaseGenerator):
```
Comment (FreakieHuang) [inline mmrazor/models/architectures/generators/dafl_generator.py:13]: UT

File: mmrazor/models/algorithms/distill/configurable/datafree_distillation.py [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+        self.generator_distiller = MODELS.build(generator_distiller)
+
+        if not isinstance(teachers, Dict):
+            raise TypeError('teacher should be a `dict` but got '
+                            f'{type(teachers)}')
+
+        self.teachers = nn.ModuleDict()
+        for teacher_name, cfg in teachers.items():
+            self.teachers[teacher_name] = MODELS.build(cfg.build_cfg)
+            if cfg.ckpt_path:
```
[the remark this answers is not in the document]
Comment (wilxy) [reply mmrazor/models/algorithms/distill/configurable/datafree_distillation.py:57]: Done

File: mmrazor/models/algorithms/distill/configurable/datafree_distillation.py [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+    # TODO: The typing of this file will report many errors in python3.7.
+    def __init__(self,
+                 distiller,
+                 generator_distiller,
+                 teachers,
+                 generator,
+                 student_iter=1,
+                 student_train_first=False,
+                 **kwargs) -> None:
+        super().__init__(**kwargs)
```
[the remark this answers is not in the document]
Comment (wilxy) [reply mmrazor/models/algorithms/distill/configurable/datafree_distillation.py:44]: Keep the same style with other Distillatoin Algorithms.

File: mmrazor/models/losses/dafl_loss.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+    def forward(self, preds_T: torch.Tensor) -> torch.Tensor:
+        """Forward function for the DAFLLoss.
+
+        Args:
+            preds_T (torch.Tensor): The predictions of teacher.
+        """
+        return self.loss_weight * self.forward_train(preds_T)
+
+    def forward_trian(self, preds_T: torch.Tensor) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (wilxy) [reply mmrazor/models/losses/dafl_loss.py:32]: Done

File: mmrazor/models/algorithms/distill/configurable/datafree_distillation.py (hunk 2 above)
Comment (wilxy) [reply mmrazor/models/algorithms/distill/configurable/datafree_distillation.py:15]: Done

File: mmrazor/models/losses/dafl_loss.py (hunk 1 above)
Comment (wilxy) [reply mmrazor/models/losses/dafl_loss.py:62]: Change its parent class to nn.Module.

File: mmrazor/models/architectures/generators/dafl_generator.py (hunk 4 above)
Comment (wilxy) [reply mmrazor/models/architectures/generators/dafl_generator.py:13]: Done

File: mmrazor/models/ops/gather_tensors.py (hunk 3 above)
Comment (wilxy) [reply mmrazor/models/ops/gather_tensors.py:8]: There are no asserts or warnings in GatherTensor and its parent except NotImplementedError. 
Add its UT in test_distillation_loss.py to check whether the default process group has been initialized.

File: mmrazor/models/algorithms/distill/configurable/datafree_distillation.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (FreakieHuang) [reply mmrazor/models/algorithms/distill/configurable/datafree_distillation.py:43]: @pppppM Then we may update other KD algo as well

File: mmrazor/models/algorithms/distill/configurable/datafree_distillation.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (FreakieHuang) [reply mmrazor/models/algorithms/distill/configurable/datafree_distillation.py]: I do not see any changes yet.

[the remark this answers is not in the document]
Comment (wilxy) [reply mmrazor/models/algorithms/distill/configurable/datafree_distillation.py]: https://github.com/open-mmlab/mmrazor/blob/254c02fe2ced755c0ba12c8982989ba6678425e0/mmrazor/models/algorithms/distill/configurable/datafree_distillation.py#L56-L60

File: mmrazor/models/architectures/generators/base_generator.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+            assert latent_data.shape[1] == self.latent_dim, \
+                'Second dimension of the input must be equal to "latent_dim",'\
+                f'but got {latent_data.shape[1]} != {self.latent_dim}.'
+            if latent_data.ndim == 2:
+                batch_data = latent_data
+            else:
+                raise ValueError('The noise should be in shape of (n, c)'
+                                 f'but got {latent_data.shape}')
+        elif latent_data is None and batch_size > 0:
+            batch_data = torch.randn((batch_size, self.latent_dim))
```
Comment (sunnyxiaohu) [inline mmrazor/models/architectures/generators/base_generator.py:52]: when latent_data is None, does `batch_size` have to be larger than 0? assert if it has to be.

Comment (wilxy) [reply mmrazor/models/architectures/generators/base_generator.py:55]: Done. Added thie assert.
https://github.com/open-mmlab/mmrazor/pull/235/files#diff-8140a05d1d793977e1ddaec0f0675306f47c9f2b268cbc63502a6bf8a55924dfR52-R54

File: mmrazor/models/algorithms/distill/configurable/datafree_distillation.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (FreakieHuang) [reply mmrazor/models/algorithms/distill/configurable/datafree_distillation.py]: Suggestion:

```python
if 'ckpt_path' in cfg:
    pass
```
